modules/core/04-channel/keeper/upgrade.go

```diff
@@ -22,115 +14,57 @@ func (k Keeper) ChanUpgradeInit(
 	ctx sdk.Context,
 	portID string,
 	channelID string,
-	chanCap *capabilitytypes.Capability,
-	proposedUpgradeChannel types.Channel,
-	counterpartyTimeoutHeight clienttypes.Height,
-	counterpartyTimeoutTimestamp uint64,
-) (upgradeSequence uint64, previousVersion string, err error) {
+	upgradeFields types.UpgradeFields,
+	upgradeTimeout types.UpgradeTimeout,
+) (types.Upgrade, error) {
 	channel, found := k.GetChannel(ctx, portID, channelID)
 	if !found {
-		return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
 	}
 
 	if channel.State != types.OPEN {
-		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
-	}
-
-	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
-		return 0, "", errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
 	}
 
-	// set the restore channel to the current channel and reassign channel state to INITUPGRADE,
-	// if the channel == proposedUpgradeChannel then fail fast as no upgradable fields have been modified.
-	restoreChannel := channel
-	channel.State = types.INITUPGRADE
-	if reflect.DeepEqual(channel, proposedUpgradeChannel) {
-		return 0, "", errorsmod.Wrap(types.ErrChannelExists, "existing channel end is identical to proposed upgrade channel end")
+	if err := k.ValidateUpgradeFields(ctx, upgradeFields, channel); err != nil {
```

Comment:
In a separate pr, how would folks feel about making this function private and placed in `upgrade.go`?

Reply:
I think I made the exported as it was used in another package, if it's possible though I'm in favour of making it unexported.

Reply:
it looks like [here](https://github.com/cosmos/ibc-go/blob/f29e2f8f0487948a36d820dd61f9542d86de91ce/modules/core/04-channel/keeper/keeper_test.go#L639)